python/its-status/its_status/__init__.py

```diff
@@ -33,4 +50,111 @@ def init(*args, **kwargs):
+                        kwargs=kwargs,
+                        daemon=True,
+                    ),
+                }
+            )
+
+
+def loop(*args, **kwargs):
+    freq = kwargs["cfg"].getfloat("generic", "frequency", fallback=1.0)
+    collect_ts = kwargs["cfg"].getboolean(
+        "generic", "timestamp_collect", fallback=False
+    )
+
+    for n in plugins["collectors"]:
+        plugins["collectors"][n]["thread"].start()
+
+    # See note about timerfd and eventfd, above.
+    timer = linuxfd.timerfd(closeOnExec=True)
+    timer.settime(value=1.0 / freq, interval=1.0 / freq)
+
+    tick = 0
+    while True:
+        try:
+            evt = timer.read()
+        except InterruptedError:
+            # Someone sent a signal to this thread...
+            continue
+
+        now = time.time()
+        tick += evt
+
+        status = dict()
+        if evt > 1:
+            status["errors"] = {
+                "timestamp": now,
+                "tick": tick,
+                "missed_ticks": evt - 1,
+            }
+            print(
+                f"tick: resuming #{tick} after {evt-1} missed ticks",
+                file=sys.stderr,
+            )
+
+        # Trigger all collectors to start collecting their data
+        for c in plugins["collectors"]:
+            os.eventfd_write(plugins["collectors"][c]["trigger_fd"], 1)
+
+        # Wait for all collectors to have finished collecting their data
+        cpt = len(plugins["collectors"])
+        while cpt:
+            # We can get release events from more than one plugin at a time,
+            # so we actually need to count them
+            cpt -= os.eventfd_read(plugins["release_fd"])
+
+        # Gather the data from all collectors
+        for c in plugins["collectors"]:
+            if c == "static":
+                status.update(plugins["collectors"][c]["data"])
+            else:
+                status[c] = plugins["collectors"][c]["data"]
+
+        if collect_ts:
+            status["collect"] = {"start": now}
+            for c in plugins["collectors"]:
+                status["collect"][c] = plugins["collectors"][c]["collect_ts"]
+            status["collect"]["duration"] = time.time() - now
+
+        # Here, we'd send them to MQTT or anywhere else
+        status["timestamp"] = time.time()
+        for e in plugins["emitters"]:
+            plugins["emitters"][e]["obj"].emit(status)
+
+
+def _plugin_loop(plugin, *args, **kwargs):
+    collect_ts = kwargs["cfg"].getboolean(
+        "generic", "timestamp_collect", fallback=False
+    )
+    while True:
+        try:
+            evt = os.eventfd_read(plugin["trigger_fd"])
+        except InterruptedError:
+            # Someone sent a signal to this thread...
+            continue
+
+        if collect_ts:
+            start = time.time()
+        plugin["data"] = plugin["obj"].capture()
+        if collect_ts:
+            plugin["collect_ts"] = {
+                "start": start,
```

Comment:
It can't possibly be not be initialised: See the line 151:
```
  151         if collect_ts:
  152             start = time.time()
  153         plugin["data"] = plugin["obj"].capture()
  154         if collect_ts:
  155             plugin["collect_ts"] = {
  156                 "start": start,
```
So, start is only ever used when collect_ts is true, and if that's the case,
it has been set a few lines above.

Reply:
Hum, dangerous for a potential refactoring further...

Reply:
I would add at least for the future debug:

```
        if collect_ts:
            start = time.time()
        else:
            start = None
```

Reply:
This does not change much: if it happens to be unset, we'd get a `NameError`
exception when we try to use it, while if we set it to _None_, we'd get a
`TypeError` exception when we try to substract a float (`time.time()`) from
it. In either case, we get an exception.

So I prefer we get the _NameError_, which will make it explicit that the variable
is not defined.